src/primitives/bezier.jl

```diff
@@ -0,0 +1,89 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENSE in the project root.
+# ------------------------------------------------------------------
+
+"""
+    BezierCurve(points)
+
+A Bézier curve with control points `points`.
+See https://en.wikipedia.org/wiki/Bézier_curve.
+A point on the curve `b` can be evaluated by calling
+`b(t)` with `t` between `0` and `1`.
+The evaluation method defaults to DeCasteljau's algorithm
+for accurate evaluation. Horner's method, faster with a
+large number of points but less precise, can be used via
+`b(t, Horner())`.
+
+## Example
+
+```julia
+BezierCurve(Point2[(0.,0.),(1.,-1.)])
+```
+"""
+struct BezierCurve{Dim,T,V<:AbstractVector{Point{Dim,T}}} <: Primitive{Dim,T}
+  controls::V
+end
+
+BezierCurve(points::Vararg) = BezierCurve(collect(points))
+BezierCurve(points::AbstractVector{<:Tuple}) = BezierCurve(Point.(points))
+
+ncontrols(b::BezierCurve) = length(b.controls)
+
+"""
+Evaluation method used to obtain a point along
+a Bézier curve from a parametric expression.
+"""
+abstract type BezierEvalMethod end
+
+"""
+Accurate evaluation using De Casteljau's algorithm.
+Recommended for a small number of control points.
+
+See https://en.wikipedia.org/wiki/De_Casteljau%27s_algorithm.
+"""
+struct DeCasteljau <: BezierEvalMethod end
+
+"""
+Approximate evaluation using Horner's method.
+Recommended for a large number of control points,
+if you can afford a precision loss.
+
+See https://en.wikipedia.org/wiki/Horner%27s_method.
+"""
+struct Horner <: BezierEvalMethod end
+
+(curve::BezierCurve)(t) = curve(t, DeCasteljau())
+
+function check_domain(::BezierCurve, t)
+  if t < 0 || t > 1
+    throw(DomainError(t, "b(t) is not defined for t outside [0, 1]."))
+  end
+end
```

Comment:
I wasn't sure whether I should define a function for this or not. The idea was not to copy paste this for the two algorithms. The first argument type is just used for dispatch. If that makes sense, I can extend this function with other methods for other types which have a restricted domain (all callable parametric types for now), in a separate PR. To have atomic changes I could just remove this here, copy-paste the error handling code, and recreate this function in a separate PR, if you deem it preferable.

Reply:
Yes, maybe we should introduce this helper function in a separate PR and then add methods to it for the different geometries like Segment, Ray and now Bezier :)